frontend/src/metabase/lib/expressions/typechecker.js

```diff
@@ -0,0 +1,88 @@
+import { ExpressionVisitor } from "./visitor";
+
+export function typeCheck(cst, startRule) {
+  const rootType = startRule || "expression";
+
+  class TypeChecker extends ExpressionVisitor {
```

Comment:
It doesn't matter, but maybe TypeChecker could have `rootType` as an argument to its constructor rather than capturing it in a closure?

I'm not really sure why that's my preference. It seems to clarify the class/calling function line.

Reply:
Very good point!